### ISSUE TITLE
Tensor subclass is not preserved during backward with gradient checkpointing

### ISSUE THREAD

**Issue by function2-llx** [pytorch/pytorch#105644]
### 🐛 Describe the bug

When creating a custom `torch.Tensor` subclass, during forward, the subclass information is properly preserved. However, when using gradient checkpointing feature, the subclass information is not kept after any calculation.

```python
import torch
from torch import nn
from torch.utils import checkpoint

class MyTensor(torch.Tensor):
    pass

class Module(nn.Linear):
    def forward(self, x):
        print('layer input type:', type(x))
        x = MyTensor(x)
        y = super().forward(x)
        print('layer output type:', type(y))
        return y

def main():
    x = MyTensor(torch.randn(1, 1))
    m1 = nn.Linear(1, 1)
    m2 = Module(1, 1)
    print('forward')
    z = checkpoint.checkpoint(m2, m1(x))
    print('output type:', type(z))
    print('backward')
    z.backward()

if __name__ == '__main__':
    main()
```

output:
```
forward
layer input type: <class '__main__.MyTensor'>
layer output type: <class '__main__.MyTensor'>
output type: <class '__main__.MyTensor'>
backward
layer input type: <class 'torch.Tensor'>
layer output type: <class 'torch.Tensor'>
```

### Versions

```
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:17) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-73-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A40
GPU 1: NVIDIA A40
GPU 2: NVIDIA A40
GPU 3: NVIDIA A40

Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          56
On-line CPU(s) list:             0-55
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 5120 CPU @ 2.20GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              14
Socket(s):                       2
Stepping:                        4
CPU max MHz:                     3200.0000
CPU min MHz:                     1000.0000
BogoMIPS:                        4400.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       896 KiB (28 instances)
L1i cache:                       896 KiB (28 instances)
L2 cache:                        28 MiB (28 instances)
L3 cache:                        38.5 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54
NUMA node1 CPU(s):               1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] numpy==1.25.1
[pip3] pytorch-lightning==2.0.5
[pip3] torch==2.0.1
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.25.1          py311h64a7726_0    conda-forge
[conda] pytorch                   2.0.1           py3.11_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-lightning         2.0.5                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchmetrics              0.11.4             pyhd8ed1ab_0    conda-forge
[conda] torchtriton               2.0.0                     py311    pytorch
[conda] torchvision               0.15.2              py311_cu118    pytorch
```

**Comment by albanD** [comment]
Hey!
It is expected that non-torch_dispatch subclass can only live on the Python side and will disappear as soon as you touch c++.
Does this clarify things: https://pytorch.org/docs/main/notes/extending.html#extending-torch-python-api ?

**Comment by function2-llx** [comment]
Hi @albanD , thank you for the information. Does this mean because gradient checkpointing is implemented in C++, if I want my custom Tensor's subclass work with gradient checkpointing, I must extend torch's native API by defining `__torch_dispatch__`?

**Comment by ezyang** [comment]
Yes, you need to use torch dispatch

**Comment by function2-llx** [comment]
Hey! After almost two years, I think I have found some hint for the actual cause for this issue, which seems to be different from the previous suggestion. In brief:
1. The current implementation of `Tensor.__torch_function__` disables `__torch_function__` for subclasses before it returns.
2. During backward, `Tensor.backward` is handled by `Tensor.__torch_function__` in the beginning, and all `__torch_function__` of subclasses are disabled. 

Let me break it down how I found this:

1. First I locate that the subclass disappeared exactly at this line after calling `Tensor.detach`:

https://github.com/pytorch/pytorch/blob/8568dbce1d7775d37c3f2dcc4073d3f44968fc43/torch/utils/checkpoint.py#L1053

[Screenshot of the callstack](https://github.com/user-attachments/assets/a77c5bda-6fc1-4f73-b90c-a3478584c0f3)

2. However, usually `Tensor.detach` should also be intercepted by `__torch_function__`. So the reason it is not here must be `__torch_function__` was disabled somewhere before. 

3. Then I traverse the callstack, and find that `__torch_function__` was disabled here:

https://github.com/pytorch/pytorch/blob/8568dbce1d7775d37c3f2dcc4073d3f44968fc43/torch/_tensor.py#L1669-L1670

And it reaches here in the beginning because of calling `Tensor.backward` on the tensor ([screenshot](https://github.com/user-attachments/assets/3ae31a63-e7ca-41bc-8720-56844f281956)). As a result, all subsequent computation during the backward will have `__torch_function__` disabled.
